File: src/main/java/brightspark/landmanager/message/MessageHomeActionKickOrPass.java
package brightspark.landmanager.message;

import brightspark.landmanager.LandManager;
import brightspark.landmanager.data.areas.Area;
import brightspark.landmanager.data.areas.CapabilityAreas;
import brightspark.landmanager.util.HomeGuiActionType;
import brightspark.landmanager.util.Utils;
import com.mojang.authlib.GameProfile;
import io.netty.buffer.ByteBuf;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.util.math.BlockPos;
import net.minecraftforge.fml.common.network.simpleimpl.IMessage;
import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler;
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext;

import java.util.UUID;

public class MessageHomeActionKickOrPass implements IMessage
{
	private BlockPos pos;
	private boolean isPass;
	private UUID uuid;

	public MessageHomeActionKickOrPass() {}

	public MessageHomeActionKickOrPass(BlockPos pos, boolean isPass, UUID uuid)
	{
		this.pos = pos;
		this.isPass = isPass;
		this.uuid = uuid;
	}

	@Override
	public void fromBytes(ByteBuf buf)
	{
		pos = BlockPos.fromLong(buf.readLong());
		isPass = buf.readBoolean();
		long most = buf.readLong();
		long least = buf.readLong();
		uuid = new UUID(most, least);
	}

	@Override
	public void toBytes(ByteBuf buf)
	{
		buf.writeLong(pos.toLong());
		buf.writeBoolean(isPass);
		buf.writeLong(uuid.getMostSignificantBits());
		buf.writeLong(uuid.getLeastSignificantBits());
	}

	public static class Handler implements IMessageHandler<MessageHomeActionKickOrPass, IMessage>
	{
		@Override
		public IMessage onMessage(MessageHomeActionKickOrPass message, MessageContext ctx)
		{
			EntityPlayerMP player = ctx.getServerHandler().player;
			CapabilityAreas cap = player.world.getCapability(LandManager.CAPABILITY_AREAS, null);
			if(cap == null)
				return null;
			Area area = cap.intersectingArea(message.pos);
			if(!Utils.canPlayerEditArea(area, player, player.world.getMinecraftServer()))
				return new MessageHomeActionReplyError("message.error.noPerm");
			UUID uuid = message.uuid;
			GameProfile profile = player.world.getMinecraftServer().getPlayerProfileCache().getProfileByUUID(uuid);
			if(profile == null)
				return new MessageHomeActionReplyError("message.error.noPlayer");

			boolean changed = true;
			if(message.isPass)
				area.setOwner(uuid);
			else
			{
				changed = area.removeMember(uuid);
				if(changed)
					cap.decreasePlayerAreasNum(uuid);
			}
			if(changed)
			{
				cap.dataChanged();
				return new MessageHomeActionReply(message.isPass ? HomeGuiActionType.PASS : HomeGuiActionType.KICK, uuid, profile.getName());
			}
			else
				return null;
		}
	}
}

File: src/main/kotlin/brightspark/landmanager/message/MessageHomeActionKickOrPass.kt
package brightspark.landmanager.message

import brightspark.ksparklib.api.Message
import brightspark.ksparklib.api.sendToPlayer
import brightspark.landmanager.LandManager
import brightspark.landmanager.data.areas.AreaUpdateType
import brightspark.landmanager.util.HomeGuiActionType
import brightspark.landmanager.util.areasCap
import brightspark.landmanager.util.canEditArea
import net.minecraft.network.PacketBuffer
import net.minecraft.util.math.BlockPos
import net.minecraftforge.fml.network.NetworkEvent
import java.util.*
import java.util.function.Supplier

class MessageHomeActionKickOrPass : Message {
	private lateinit var pos: BlockPos
	private var isPass: Boolean = false
	private lateinit var uuid: UUID

	@Suppress("unused")
	constructor()

	constructor(pos: BlockPos, isPass: Boolean, uuid: UUID) {
		this.pos = pos
		this.isPass = isPass
		this.uuid = uuid
	}

	override fun encode(buffer: PacketBuffer): Unit = buffer.run {
		writeBlockPos(pos)
		writeBoolean(isPass)
		writeUniqueId(uuid)
	}

	override fun decode(buffer: PacketBuffer): Unit = buffer.run {
		pos = readBlockPos()
		isPass = readBoolean()
		uuid = readUniqueId()
	}

	override fun consume(context: Supplier<NetworkEvent.Context>) {
		context.get().enqueueWork {
			val player = context.get().sender ?: return@enqueueWork
			val world = player.world
			val server = world.server ?: return@enqueueWork
			val cap = world.areasCap
			val area = cap.intersectingArea(pos)
			if (!player.canEditArea(area)) {
				LandManager.NETWORK.sendToPlayer(MessageHomeActionReplyError("message.landmanager.error.noPerm"), player)
				return@enqueueWork
			}
			val profile = server.playerProfileCache.getProfileByUUID(uuid) ?: run {
				LandManager.NETWORK.sendToPlayer(MessageHomeActionReplyError("message.landmanager.error.noPlayer"), player)
				return@enqueueWork
			}

			var changed = true
			if (isPass) {
				area!!.owner?.let { area.addMember(it) }
				area.owner = uuid
				area.removeMember(uuid)
			} else {
				changed = area!!.removeMember(uuid)
				if (changed)
					cap.decreasePlayerAreasNum(uuid)
			}
			if (changed) {
				cap.dataChanged(area, AreaUpdateType.CHANGE)
				LandManager.NETWORK.sendToPlayer(MessageHomeActionReply(if (isPass) HomeGuiActionType.PASS else HomeGuiActionType.KICK, uuid, profile.name), player)
			}
		}
	}
}
